proposals/lambdas-06-decision-tree.md

```diff
@@ -0,0 +1,21 @@
+So now, knowing those type relationships, we can start to develop some best practices which will help people navigate which types of blocks/lambdas to use in their program architectures:
+
+In general when designing code that accepts a block/lambda for use, you want to give maximum freedom to the calling code that is providing the block/lambda while still meeting your own needs for how you will use it.
+
+With that in mind, here is a initial decision tree to use as an API design convention:
+
+1. "I will only call the lambda immediately within the function that receives it"
+  - if true: use `yield` and let the compiler infer the optimal underlying type based on yield sites, and let it have the best performance optimizations
+  - if false: (continue below)
+
+2. "I can guarantee that I will only call this function once at most"
+  - if true: use `FunOnce` (a.k.a `FunOnce'iso`)
+  - if false: (continue below)
+
+3. "This lambda will never need to be taken across an actor/region boundary (i.e. it doesn't need to be sendable)"
+  - if true: use `Fun` (a.k.a `Fun'ref`)
+  - if false: (continue below)
+
+4. "All of the arguments I need to pass into this lambda are sendable and the result I get out from it can also be sendable"
+  - if true: use `Fun'iso`
+  - if false: use `FunVal` (a.k.a `FunVal'val`)
```

Comment:
Do we have any use case where taking a `box` fun could be useful for callers?

Reply:
Ah no that would be guaranteeing that the best that we can give is box, so we'd have to have both refs and vals which is an enormous mess I could never see happening with lambdas

Reply:
I guess if it gets stored such that it never leaves the thread but may need to be called from a box method...

Reply:
I suppose this is a potential use case, though I'm not sure how common it will be. I was hoping to avoid a fourth trait to avoid adding more mental load on which trait to use, but it seems this one may not be totally avoidable.

I didn't really want to consider it because it is strictly weaker at the block site and I didn't see it as having any benefits. But you're right - it is beneficial at the lambda object use site for the specific case where the lambda object user needs to store the lambda in a field and then retrieve and call it from that field in the middle of a `:fun box`, where it cannot retrieve and call a `:fun ref` on that field.

If I add nothing beyond what is already in the proposal, in such a situation the lambda user would need to specify `FunVal'val`, meaning the lambda supplier would be unnecessarily restricted in not being able to capture non-shareable values from the outer scope. I say it's unnecessary because we've established in this example that the lambda user doesn't actually need it to be sendable, so `val` is unnecessarily strong where `box` could be used.

---

So we need a lambda "Kind" that has the following properties:

Call count:
  - can be called zero or more times

Inter-scope interaction:
  - can capture variables from the outer scope
  - CANNOT rebind captured variables in the inner scope
  - CANNOT consume `iso` variables in the inner scope
  - CANNOT affect variable bindings in the outer scope
  - CANNOT raise an error into the outer scope
  - CANNOT return early, exiting the outer scope

Ref cap restrictions:
  - can capture non-shareable values from the outer scope
  - HOWEVER, all captured values are seen through a `box` lens
  - can receive non-sendable yielded values
  - can return a non-sendable yield result value

Mobility:
  - CANNOT cross an actor boundary
  - can be objectified and stored away for later

Note that the only difference from the `FunVal'val` kind is that it is able to capture non-shareable values from the outer scope. However, all the captured values would be seen through a `box` lens, which is unfortunate but necessary.

---

Perhaps we can do this without adding a fourth trait?

Maybe we could amend the third trait, which is currently defined as:

```savi
:trait val FunVal(R, P1, P2)
  :fun val call(P1, P2) R
```

and we amend it such that it looks like this:

```savi
:trait val FunRead(R, P1, P2)
  :fun box call(P1, P2) R
```

So be default `FunRead` would be `FunRead'val`, but in this less common kind of case we are discussing `FunRead'box` could potentially be used, which would behave like what we have discussed above.

I think that works, without us needing to add the cognitive load of a fourth trait.

What do you think?

Reply:
I agree, I think it's incredibly rare that FunVal is needed, but it could be possible, if the function is sending not just its arguments, but itself as a whole. This just can't happen with lambdas, and for non-lambdas a somewhat cumbersome version may be able to store a val reference to itself

Reply:
By agree I mean that `FunRead` defaulting to val and being the interface would be a good move.

Then we have:

Once -> FunOnce
Mutation available -> FunMut'ref in one actor, or FunMut'iso for sending
Mutation unavailable -> FunRead'box in one actor, or FunRead'val for sharing

Reply:
Yes, I think we've arrived at a sane hierarchy for the purpose of supporting lambdas.

Thanks for your help!

I will revise this PR and the associated comments on the other ticket